### ISSUE TITLE
Term Policy `verbatim::` doesn't appear to work at all, nor does `qos::` on Juniper

### ISSUE THREAD

**Issue by bdlamprecht** [google/capirca#175]
According the the documentation [here](https://github.com/google/capirca/blob/master/doc/policy_format.txt#L138), the term policy `verbatim` **must** be supported.

But when I try to test it out as in the following code, I get the error that follows:
[in file `testing_ios_junos.pol`]
```
header {
  comment:: "/----------------------------------------------"
  comment:: " QoS class 1 - IP address classification policy"
  comment:: "----------------------------------------------/"
  target:: juniper cm_class1_ipaddr inet
  target:: cisco al_class1_ipaddr extended
}

term al_att_class1_ipaddr_seq_100 {
  source-address:: 9DOT
  source-port:: UPPER_HALF
  destination-address:: HCS_AMERICAS
  destination-port:: UPPER_HALF
  protocol:: udp
  action:: accept
  loss-priority:: low
  counter:: al_class1_ipaddr_seq_100
  verbatim: "THIS IS A TEST"
}
```
This is the error that I get:
```
$ aclgen
I0802 12:57:00.588260 140023978403584 aclgen.py:491] finding policies...
/home/bl839s/.virtualenvs/caprica/lib/python3.6/site-packages/ply-3.11-py3.6.egg/ply/yacc.py:183: UserWarning: PLY: Don't use global functions errok(), token(), and restart() in p_error().
Instead, invoke the methods on the associated parser instance:

    def p_error(p):
        ...
        # Use parser.errok(), parser.token(), parser.restart()
        ...

    parser = yacc.yacc()

  warnings.warn(_warnmsg)
W0802 12:57:00.707302 140023978403584 aclgen.py:513]

error encountered in rendering process:
Error parsing policy file ./policies/pol/testing_ios_junos.pol:
<class 'capirca.lib.policy.ParseError'> ERROR on ""THIS IS A TEST"" (type DQUOTEDSTRING, line 18, Next '}')


I0802 12:57:00.708703 140023978403584 aclgen.py:444] no files changed, not writing to disk
W0802 12:57:00.708981 140023978403584 aclgen.py:519] done, with errors.
```
---

Similarity, for Juniper devices, when I try and use the term `qos::` (which is supposed to be supported on Juniper as noted [here](https://github.com/google/capirca/blob/master/doc/policy_format.txt#L152) and silently ignored on other platforms), such as in the continued test under the same `header`:
```
term al_class1_ipaddr_seq_110 {
  source-address:: 9DOT
  destination-address:: HCS_AMERICAS
  destination-port:: REAL_TIME_MEDIA
  protocol:: udp
  action:: accept
  loss-priority:: low
  counter:: al_class1_ipaddr_seq_110
  qos:: ef
}
```

```
---SNIP---
error encountered in rendering process:
Error parsing policy file ./policies/pol/testing_ios_junos.pol:
<class 'capirca.lib.policy.ParseError'> ERROR on "ef" (type DSCP, line 29, Next '}')
---SNIP---
```

When either of those lines (`verbatim::` or `qos::`) are commented out, everything seems to work.  Please let me know if I'm doing something incorrectly or if this is something to do with the conversion from python2 to python3 (issue #154).

**Comment by ankenyr** [comment]
Reading the tests could likely help in understanding how these work.
https://github.com/google/capirca/blob/375bed4a957098912daf9364293be3d467149194/tests/lib/ciscoasa_test.py
Terms can be used in in headers that have multiple platforms or in includes that are used across policies that have multiple platforms. Verbatim is printed out by a generator without any modifications. Due to the two things above it would be necessary to only print out verbatim on the correct platform or else it would limit its use.

As for QOS I think you are looking for ef1

**Comment by bdlamprecht** [comment]
Okay, that makes sense regarding the `verbatim` keyword.
I tried adding the keyword `verbatim` within another `term` and was given this error:
```
error encountered in rendering process:
Error parsing policy file ./policies/pol/testing_ios_junos.pol:
<class 'capirca.lib.policy.ParseError'>term "al_att_class1_ipaddr_seq_100" has both verbatim and non-verbatim tokens.
```

**Comment by ankenyr** [comment]
The error message seems to explain the problem. The code is a bit more descriptive.
https://github.com/google/capirca/blob/master/capirca/lib/policy.py#L1176
Using verbatim along with some core keywords ends up causing an error.

What are you trying to do using verbatim? Would that be better to have as a keyword that is available for use within Capirca?

**Comment by bdlamprecht** [comment]
Sorry, I was finished typing my entire comment before I accidentally submitted it (`ctrl+enter` works differently on GitHub than it does on Slack).

I finished it and saved it above :smile:

**Comment by ankenyr** [comment]
I am not sure what you want with regards to examples here. Are you asking for an example in the policies folder that specifically includes a verbatim?
That already exists
https://github.com/google/capirca/blob/master/policies/pol/sample_speedway.pol#L35

No worries on the QOS, I was confused about that myself as I still am not sure why 2 letters isn't accepted. I think the parsers RE does that but I didn't digest it enough to know 100%

Not 100% sure about your last question
The code says
`interface_specific = 'not-interface-specific' not in filter_options[1:]`
It is used in a couple of places but eventually for rendering it is used here
https://github.com/google/capirca/blob/375bed4a957098912daf9364293be3d467149194/capirca/lib/juniper.py#L922

Not sure if that helps but please let me know if it doesn't.

And no worries, I do the opposite and post too soon on slack all the time.

**Comment by bdlamprecht** [comment]
In regards to the `interface-specific`, my question was how to remove it as the default appears to have it included (lack of documented examples as as in something like `readthedocs.io`).

**Comment by ankenyr** [comment]
I try to be active in there, not sure how I missed it since I should get alerts to anyone saying "Capirca" or posting in the channel.

Those ID's are p4 tags. They are also referenced as such in the code
https://github.com/google/capirca/blob/master/capirca/aclgen.py#L379
These are used in perforce, not really useful unless you use perforce as a source control system.

What would the difference between readthedocs.io and the wiki page on github be? We already have things documented in there. We also have samples included here
https://github.com/google/capirca/tree/master/policies
Understanding what the actual issue is here would help. I am not sure if you are saying there is no documentation/examples, or that there is a lack of examples.

Adding "not-interface-specific" to the options would remove it.
https://github.com/google/capirca/wiki/Policy-format#juniper
In other words change this
https://github.com/google/capirca/blob/master/policies/pol/sample_multitarget.pol#L10
to
`target:: juniper edge-inbound inet not-interface-specific`

**Comment by bdlamprecht** [comment]
Yeah, no worries.  I just hate to have a conversation here on GitHub when other alternatives exist.
Regarding the benefit of using `readthedocs` vs the `wiki` here, it is much easier to search or things on `readthedocs` when you don't know exactly what you're looking for and where it might be at.

For example, searching in the `wiki` for the term `specific` results in no results.  Where as searching for `webhooks` on `netbox.readthedocs.io` results in various results of where you might be able to find what you're looking for.

I know this is minor, but it would help for complete newbie's to `caprica` such as myself.

---
Regarding this comment:
>Those ID's are p4 tags. They are also referenced as such in the code

Would it be possible to use those for some type of `git` tags?  I'm thinking such as mapping the `$Date$` to the date of a commit and `$Revision$` to the commit log.

**Comment by bdlamprecht** [comment]
Also, the link on the `wiki` for "Aaron's Quick Intro to Google’s Capirca" doesn't appear to be active anymore.

**Comment by ankenyr** [comment]
Well there are pro's and con's to each thing. Slack is great for quick questions and I use it for that. For reporting of issues or feature requests I expect people to use Github. There is a maintained history in Github and it allows people to see others are experiencing an issue or that some problem exists or that there is demand for a feature. Slack does not give us that.

Not sure about searching being easier on readthedocs.io. I can search for verbatim in the wiki here.
https://github.com/google/capirca/search?q=verbatim&type=Wikis
Everyone has their own personal way they would like things. Some would love a classic man page or flat text file, some would want the wiki page here on github, others may want alternatives but something similar to readthedocs.io. Some insane people believe the code and its docstrings are the documentation.

I hope this doesn't sound rude or cold but considering we will never satisfy everyone I would not be entertaining the idea of moving to a different platform for documentation. If documentation is lacking in some way I would be happy to address it but there is a balance to be struck here. How many examples do we need to provide? An exhaustive list of every possible combination? Some subset? At what point do we stop writing documentation? What is the point of error messages in this case if they are not telling you that you are doing something wrong?

You are right, Aaron's blog is down. I will remove that link.

I am going to close this issue as the QOS question and verbatim one are resolved. If you want to file a feature request about the Github stuff please file one. It could be interesting.
If you have ideas on improvements to the documentation please feel free to suggest specific cases we could improve. You can also I believe clone the wikipedia repo and send PR's suggesting changes.
